### PR TITLE
Add comments to translation interfaces, clarify on error return value cannot be used

### PR DESCRIPTION
**Pull request by bogdandrutu**
Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

**Review by carlosalberto — APPROVED** [review]

File: internal/model/decoder.go [hunk 1]
```diff
@@ -17,17 +17,20 @@ package model
 // MetricsDecoder interface to decode bytes into protocol-specific data model.
 type MetricsDecoder interface {
 	// DecodeMetrics decodes bytes into protocol-specific data model.
+	// If error is not nil returned interface cannot be used.
```
Comment (jpkrohling) [inline internal/model/decoder.go:20]:
```suggestion
	// If the returned error is not nil, the returned interface cannot be used.
```

I wonder if implementation shouldn't be required to return nil for the interfaces in this case?

File: internal/model/decoder.go [hunk 2]
```diff
@@ -17,17 +17,20 @@ package model
...
 	DecodeMetrics(buf []byte) (interface{}, error)
 }
 
 // TracesDecoder interface to decode bytes into protocol-specific data model.
 type TracesDecoder interface {
 	// DecodeTraces decodes bytes into protocol-specific data model.
+	// If error is not nil returned interface cannot be used.
```
Comment (jpkrohling) [inline internal/model/decoder.go:27]:
```suggestion
	// If the returned error is not nil, the returned interface cannot be used.
```

File: internal/model/decoder.go [hunk 3]
```diff
@@ -17,17 +17,20 @@ package model
...
 	DecodeTraces(buf []byte) (interface{}, error)
 }
 
 // LogsDecoder interface to decode bytes into protocol-specific data model.
 type LogsDecoder interface {
 	// DecodeLogs decodes bytes into protocol-specific data model.
+	// If error is not nil returned interface cannot be used.
```
Comment (jpkrohling) [inline internal/model/decoder.go:34]:
```suggestion
	// If the returned error is not nil, the returned interface cannot be used.
```

File: internal/model/encoder.go [hunk 4]
```diff
@@ -17,17 +17,20 @@ package model
 // MetricsEncoder encodes protocol-specific data model into bytes.
 type MetricsEncoder interface {
 	// EncodeMetrics converts protocol-specific data model into bytes.
+	// If error is not nil returned bytes slice cannot be used.
```
Comment (jpkrohling) [inline internal/model/encoder.go:20]:
```suggestion
	// If the error is not nil, the returned byte slice cannot be used.
```
Same as the case mentioned before: shouldn't the slice be nil then?

File: internal/model/encoder.go [hunk 5]
```diff
@@ -17,17 +17,20 @@ package model
...
 	EncodeMetrics(model interface{}) ([]byte, error)
 }
 
 // TracesEncoder encodes protocol-specific data model into bytes.
 type TracesEncoder interface {
 	// EncodeTraces converts protocol-specific data model into bytes.
+	// If error is not nil returned bytes slice cannot be used.
```
Comment (jpkrohling) [inline internal/model/encoder.go:27]:
```suggestion
	// If the error is not nil, the returned byte slice cannot be used.
```

File: internal/model/encoder.go [hunk 6]
```diff
@@ -17,17 +17,20 @@ package model
...
 	EncodeTraces(model interface{}) ([]byte, error)
 }
 
 // LogsEncoder encodes protocol-specific data model into bytes.
 type LogsEncoder interface {
 	// EncodeLogs converts protocol-specific data model into bytes.
+	// If error is not nil returned bytes slice cannot be used.
```
Comment (jpkrohling) [inline internal/model/encoder.go:34]:
```suggestion
	// If the error is not nil, the returned byte slice cannot be used.
```

File: internal/model/from_translator.go [hunk 7]
```diff
@@ -19,17 +19,20 @@ import "go.opentelemetry.io/collector/consumer/pdata"
 // FromMetricsTranslator interface to translate pdata.Metrics to protocol-specific data model.
 type FromMetricsTranslator interface {
 	// FromMetrics translates pdata.Metrics to protocol-specific data model.
+	// If error is not nil returned pdata.Metrics cannot be used.
```
Comment (jpkrohling) [inline internal/model/from_translator.go:22]:
```suggestion
	// If the error is not nil, the returned pdata.Metrics cannot be used.
```

File: internal/model/from_translator.go [hunk 8]
```diff
@@ -19,17 +19,20 @@ import "go.opentelemetry.io/collector/consumer/pdata"
...
 	FromMetrics(md pdata.Metrics) (interface{}, error)
 }
 
 // FromTracesTranslator interface to translate pdata.Traces to protocol-specific data model.
 type FromTracesTranslator interface {
 	// FromTraces translates pdata.Traces to protocol-specific data model.
+	// If error is not nil returned pdata.Traces cannot be used.
```
Comment (jpkrohling) [inline internal/model/from_translator.go:29]:
```suggestion
	// If the error is not nil, the returned pdata.Traces cannot be used.
```

File: internal/model/from_translator.go [hunk 9]
```diff
@@ -19,17 +19,20 @@ import "go.opentelemetry.io/collector/consumer/pdata"
...
 	FromTraces(td pdata.Traces) (interface{}, error)
 }
 
 // FromLogsTranslator interface to translate pdata.Logs to protocol-specific data model.
 type FromLogsTranslator interface {
 	// FromLogs translates pdata.Logs to protocol-specific data model.
+	// If error is not nil returned pdata.Logs cannot be used.
```
Comment (jpkrohling) [inline internal/model/from_translator.go:36]:
```suggestion
	// If the error is not nil, the returned pdata.Logs cannot be used.
```

**Review by jpkrohling — COMMENTED** [review]
The "If error" comments could have a couple of articles to make the grammar checks happier :-) Also, the "nil" comments made me confused for a while: should I expect implementations to return nil, or may they return invalid values? Some of the changes made me believe that they are nil, but the new comments made me unsure.

File: internal/model/marshaler.go [hunk 10]
```diff
@@ -22,6 +22,8 @@ import (
 
 // TracesMarshaler marshals pdata.Traces into bytes.
 type TracesMarshaler interface {
+	// Marshal the given pdata.Traces as bytes.
```
Comment (rakyll) [inline internal/model/marshaler.go:25]: Nit: Marshal into

Marshal the given pdata.Traces into bytes.

File: internal/model/marshaler.go [hunk 11]
```diff
@@ -53,6 +55,8 @@ func (t *tracesMarshaler) Marshal(td pdata.Traces) ([]byte, error) {
 
 // MetricsMarshaler marshals pdata.Metrics into bytes.
 type MetricsMarshaler interface {
+	// Marshal the given pdata.Metrics as bytes.
```
Comment (rakyll) [inline internal/model/marshaler.go:58]: Ditto.

File: internal/model/marshaler.go [hunk 12]
```diff
@@ -84,6 +88,8 @@ func (t *metricsMarshaler) Marshal(td pdata.Metrics) ([]byte, error) {
 
 // LogsMarshaler marshals pdata.Logs into bytes.
 type LogsMarshaler interface {
+	// Marshal the given pdata.Logs as bytes.
```
Comment (rakyll) [inline internal/model/marshaler.go:91]: Ditto.

File: internal/model/to_translator.go [hunk 13]
```diff
@@ -16,17 +16,23 @@ package model
 
 import "go.opentelemetry.io/collector/consumer/pdata"
 
+// ToMetricsTranslator interface to translate pdata.Metrics to protocol-specific data model.
```
Comment (rakyll) [inline internal/model/to_translator.go:19]: ToMetricsTranslator is an interface...

File: internal/model/to_translator.go [hunk 14]
```diff
@@ -16,17 +16,23 @@ package model
...
 type ToMetricsTranslator interface {
-	// ToMetrics converts a protocol-specific data model into pdata.
+	// ToMetrics converts a protocol-specific data model into pdata.Metrics.
+	// If error is not nil returned pdata.Metrics cannot be used.
 	ToMetrics(src interface{}) (pdata.Metrics, error)
 }
 
+// ToTracesTranslator interface to translate pdata.Traces to protocol-specific data model.
```
Comment (rakyll) [inline internal/model/to_translator.go:26]: Ditto.

File: internal/model/to_translator.go [hunk 15]
```diff
@@ -16,17 +16,23 @@ package model
...
 type ToTracesTranslator interface {
-	// ToTraces converts a protocol-specific data model into pdata.
+	// ToTraces converts a protocol-specific data model into pdata.Traces.
+	// If error is not nil returned pdata.Traces cannot be used.
 	ToTraces(src interface{}) (pdata.Traces, error)
 }
 
+// ToLogsTranslator interface to translate pdata.Logs to protocol-specific data model.
```
Comment (rakyll) [inline internal/model/to_translator.go:33]: Ditto.

File: internal/model/to_translator.go (hunk 13 above)
Comment (rakyll) [reply internal/model/to_translator.go:19]: This interface names could have been just "MetricsTranslator".

**Review by rakyll — APPROVED** [review]

**Review by tigrannajaryan — APPROVED** [review]

File: internal/model/marshaler.go (hunk 10 above)
Comment (bogdandrutu) [reply internal/model/marshaler.go]: Done.

File: internal/model/marshaler.go (hunk 11 above)
Comment (bogdandrutu) [reply internal/model/marshaler.go]: Done.

File: internal/model/marshaler.go (hunk 12 above)
Comment (bogdandrutu) [reply internal/model/marshaler.go]: Done.

File: internal/model/to_translator.go (hunk 14 above)
Comment (bogdandrutu) [reply internal/model/to_translator.go]: Done.

File: internal/model/to_translator.go (hunk 15 above)
Comment (bogdandrutu) [reply internal/model/to_translator.go]: Done.

File: internal/model/to_translator.go (hunk 13 above)
Comment (bogdandrutu) [reply internal/model/to_translator.go]: Done. Name cannot be changed because we have a FromMetricsTranslator as well.

**Pull request merged**